buildSrc/src/main/kotlin/org/jetbrains/DokkaBuildProperties.kt

```diff
@@ -0,0 +1,53 @@
+package org.jetbrains
+
+import org.gradle.api.provider.Provider
+import org.gradle.api.provider.ProviderFactory
+import org.gradle.jvm.toolchain.JavaLanguageVersion
+import org.jetbrains.kotlin.gradle.dsl.KotlinVersion
+import javax.inject.Inject
+
+/**
+ * Common build properties used to build Dokka subprojects.
+ *
+ * This is an extension created by the [org.jetbrains.conventions.Base_gradle] convention plugin.
+ *
+ * Default values are set in the root `gradle.properties`, and can be overridden via
+ * [CLI args, system properties, and environment variables](https://docs.gradle.org/current/userguide/build_environment.html#sec:project_properties)
+ */
+abstract class DokkaBuildProperties @Inject constructor(
+    private val providers: ProviderFactory,
+) {
+
+    /**
+     * The main version of Java that should be used to build Dokka source code.
+     *
+     * Updating the Java target is a breaking change.
+     */
+    val mainJavaVersion: Provider<JavaLanguageVersion> =
+        dokkaProperty("javaToolchain.mainCompiler") { JavaLanguageVersion.of(it) }
+
+    /**
+     * The version of Java that should be used to run Dokka tests.
+     *
+     * This value is set in CI/CD environments to make sure that Dokka still works with different
+     * versions of Java.
+     */
+    val testJavaLauncherVersion: Provider<JavaLanguageVersion> =
+        dokkaProperty("javaToolchain.testLauncher") { JavaLanguageVersion.of(it) }
+
+    /**
+     * The Kotlin language level that Dokka artifacts are compiled to support.
+     *
+     * Updating the language level is a breaking change.
+     */
+    val kotlinLanguageLevel: Provider<KotlinVersion> =
+        dokkaProperty("kotlinLanguageLevel") { KotlinVersion.fromVersion(it) }
+
+
+    private fun <T : Any> dokkaProperty(name: String, convert: (String) -> T) =
+        providers.gradleProperty("org.jetbrains.dokka.$name").map(convert)
```

Comment:
```suggestion
    private fun <T : Any> dokkaProperty(name: String, transformer: (String?) -> T) =
        providers.gradleProperty("org.jetbrains.dokka.$name").map(transformer)
```

Reply:
`String` is nullable here?

Reply:
No, it's not. If the Gradle property doesn't exist, then the map doesn't get called. 

For example, if `does-not-exist` is not defined, then the `map {}` does not get called.

```kotlin
val prop = providers.gradleProperty("does-not-exist").map { error("failed to map $it") } 

println("prop is ${prop.orNull}") // Output: prop is null
println("prop is ${prop.get()}")  // Error: Cannot query the value of this provider because it has no value available.
```

Reply:
Gotcha!